File: Sort/Intersection of Two Arrays II/IntersectionofTwoArraysII.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  var obj = {}
  var res = []
  nums1.forEach(v => obj[v] ? obj[v] += 1 : obj[v] = 1)
  nums2.forEach(v => obj[v] && obj[v]-- > 0 ? res.push(v) : null)
  return res
};

File: Sort/Intersection of Two Arrays II/IntersectionofTwoArraysII.cpp
class Solution
{
  public:
    vector<int> intersect(vector<int> &nums1, vector<int> &nums2)
    {
        unordered_map<int, int> m;
        vector<int> r;

        for (int i = 0; i < nums1.size(); i++)
        {
            int num = nums1[i];
            if (m.find(num) == m.end())
            {
                m[num] = 1;
            }
            else
            {
                m[num] += 1;
            }
        }

        for (int i = 0; i < nums2.size(); i++)
        {
            int num = nums2[i];
            if (m.find(num) != m.end() && m[num] != 0)
            {
                r.push_back(num);
                m[num] -= 1;
            }
        }

        return r;
    }
};